### PR TITLE
[onert] Implement Range Operation in cpu backend

### PR DESCRIPTION
**Pull request by intom**
Draft : https://github.com/Samsung/ONE/pull/1175

- Support RANGE Operation

Signed-off-by: sungho22.lee <sungho22.lee@samsung.com>

### CODE REVIEW

**Comment by intom** [comment]
@nnfw-bot test format-checker

**Comment by intom** [comment]
If I add the below in ONE/tests/nnapi/nnapi_gtest.skip.x86_64-linux.cpu, x86-64 test is success.
GeneratedTests.zeros_like_ex_2D_float
GeneratedTests.zeros_like_ex_4D_int32
but the current PR is failed.
I need your help.

**Comment by hseok-oh** [comment]
```diff
--- a/runtime/onert/backend/cpu/KernelGenerator.cc
+++ b/runtime/onert/backend/cpu/KernelGenerator.cc
@@ -1167,6 +1167,8 @@ void KernelGenerator::visit(const ir::operation::ZerosLike &node)
   auto fn = std::make_unique<::onert::backend::cpu::kernel::ZerosLikeLayer>();
 
   fn->configure(input_alloc, output_alloc);
+
+  _return_fn = std::move(fn);
 }
```

` _return_fn = std::move(fn);` is removed in `ZerosLikeLayer`'s `KernelGenerator` when you copy code

**Comment by intom** [comment]
> ```diff
> --- a/runtime/onert/backend/cpu/KernelGenerator.cc
> +++ b/runtime/onert/backend/cpu/KernelGenerator.cc
> @@ -1167,6 +1167,8 @@ void KernelGenerator::visit(const ir::operation::ZerosLike &node)
>    auto fn = std::make_unique<::onert::backend::cpu::kernel::ZerosLikeLayer>();
>  
>    fn->configure(input_alloc, output_alloc);
> +
> +  _return_fn = std::move(fn);
>  }
> ```
> 
> ` _return_fn = std::move(fn);` is removed in `ZerosLikeLayer`'s `KernelGenerator` when you copy code

Thank you so much!

File: runtime/onert/core/src/compiler/OperationValidator.cc [hunk 1]
```diff
@@ -1292,5 +1292,17 @@ void OperationValidator::visit(const ir::operation::LogicalOr &node)
...
+{
+  const auto output_index{node.getOutputs().at(0)};
+  const auto start_index{node.getInputs().at(ir::operation::Range::Input::START)};
+  const auto limit_index{node.getInputs().at(ir::operation::Range::Input::LIMIT)};
+  const auto delta_index{node.getInputs().at(ir::operation::Range::Input::DELTA)};
+
+  UNUSED_RELEASE(output_index);
+  UNUSED_RELEASE(start_index);
+  UNUSED_RELEASE(limit_index);
+  UNUSED_RELEASE(delta_index);
```
Comment (glistening) [inline runtime/onert/core/src/compiler/OperationValidator.cc:1305]: @intom There are no validation code at all. Do you write this code in advance for someone who may add actual validation?

File: runtime/onert/backend/cpu/kernel/RangeLayer.cc [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+void RangeLayer::run()
+{
+  switch (_output->data_type())
+  {
+    case OperandType::FLOAT32:
+      nnfw::cker::Range<float *>(reinterpret_cast<float *>(_start->buffer()),
+                                 reinterpret_cast<float *>(_limit->buffer()),
+                                 reinterpret_cast<float *>(_delta->buffer()),
+                                 reinterpret_cast<float *>(_output->buffer()));
```
Comment (glistening) [inline runtime/onert/backend/cpu/kernel/RangeLayer.cc:53]: @intom `tflite` expects and checks `start`, `limit` and `delta` is scalar in `Prepare()`. However, I cannot find check in either loader, kernel generator, op validator. Please add check in somewhere.

File: runtime/onert/backend/cpu/kernel/RangeLayer.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+public:
+  RangeLayer();
+
+  void configure(const operand::Tensor *start, const operand::Tensor *limit,
+                 const operand::Tensor *delta, operand::Tensor *output);
+
+  void run();
+  void runSync() { run(); }
+
+private:
```
Comment (dayo09) [inline runtime/onert/backend/cpu/kernel/RangeLayer.h:43]: you can delete duplicated `private`

**Comment by intom** [comment]
> Find my comments, and please update NeuralNetworksEx.h as I wrote in [#1175 (comment)](https://github.com/Samsung/ONE/pull/1175#issuecomment-631798416).

It looks like that there is already ANEURALNETWORKS_RANGE_EX in NeuralNetworksEx.h

**Comment by glistening** [comment]
@intom Again, please read my comment https://github.com/Samsung/ONE/pull/1175#issuecomment-631798416.

> Please add comment about `RANGE` in `NeuralNetworksEx.h`.

All operator needs description about its input, output and what it performs.

File: runtime/onert/backend/cpu/kernel/RangeLayer.cc (hunk 2 above)
Comment (intom) [reply runtime/onert/backend/cpu/kernel/RangeLayer.cc:53]: > @intom `tflite` expects and checks `start`, `limit` and `delta` are scalar in `Prepare()`. However, I cannot find check in either loader, kernel generator, op validator. Please add check in somewhere.

I will add the check code in runtime/onert/core/src/compiler/OperationValidator.cc
Is it enough to add it in operation validator?

File: runtime/onert/backend/cpu/kernel/RangeLayer.h (hunk 3 above)
Comment (intom) [reply runtime/onert/backend/cpu/kernel/RangeLayer.h:43]: I will delete.

File: runtime/onert/core/src/compiler/OperationValidator.cc (hunk 1 above)
Comment (intom) [reply runtime/onert/core/src/compiler/OperationValidator.cc:1305]: I will remove and add the dimension check code.

**Comment by intom** [comment]
> @intom Again, please read my comment [#1175 (comment)](https://github.com/Samsung/ONE/pull/1175#issuecomment-631798416).
> 
> > Please add comment about `RANGE` in `NeuralNetworksEx.h`.
> 
> All operator needs description about its input, output and what it performs.

I see what you mean. Thanks

**Comment by intom** [comment]
> format-checker failed, please fix format by following commands (need clang-format 3.9):
> 
> * `./nnas format`
> * `git commit -a -s --amend`
> 
> Or apply [format.patch](http://nnfw.mooo.com:8080/job/nnfw/job/master/job/pr-format-checker/2159/artifact/format.patch) using following commands:
> 
> * `git apply format.patch`
> * `git commit -a -s --amend`
> 
> Or there may be files having CRLF as newline. Please use LF.
> 
> To re-run this test, leave a comment `@nnfw-bot test format-checker`

@nnfw-bot test format-checker

**Comment by hseok-oh** [comment]
@intom Please resolve conflit.

**Comment by intom** [comment]
> @intom Please resolve conflict.

I rebased

**Comment by hseok-oh** [comment]
@dayo09 PTAL

**Pull request merged**